torchdynamo/eval_frame.py

```diff
@@ -524,6 +555,66 @@ def output(self, target, args, kwargs):
         graph,
     ).transform()
 
+    if guard_args:
+        id_to_arg_mapping = dict()
+        for i in range(len(args)):
+            a = args[i]
+            id_to_arg_mapping[id(a)] = (i, a)
+        placeholders = [n for n in new_graph.graph.nodes if n.op == "placeholder"]
+        if len(placeholders) > 0:
+            for guard in out_guards:
+                if guard.guard_types and "TENSOR_MATCH" in guard.guard_types:
+                    ref = guard.obj_weakref
+                    assert ref is not None, "TENSOR_MATCH guard object must be valid"
+                    underlying_tensor_id = id(ref())
+                    if underlying_tensor_id in id_to_arg_mapping.keys():
+                        arg_for_guard_position = id_to_arg_mapping[
+                            underlying_tensor_id
+                        ][0]
+                        arg_for_guard = id_to_arg_mapping[underlying_tensor_id][1]
+
+                        with new_graph.graph.inserting_after(placeholders[-1]):
+                            shape = (
+                                None
+                                if torchdynamo.config.dynamic_shapes
+                                else arg_for_guard.shape
+                            )
+                            strides = (
+                                None
+                                if torchdynamo.config.dynamic_shapes
```

Comment:
If `dynamic_shapes` is true doesn't `guard_args` just do nothing?

Reply:
The pydoc is correct, my impl is wrong ;)

> If torchdynamo is in dynamic_shapes mode, we assert that the dtypes of (tensor) input values match subsequent invocations.

dtype should always be checked. Good catch.